Feat/mcxa276 hal

bogdan-petru opened the pull request:
## Summary

Initial DMA driver implementation for MCXA276 with Embassy-style async API.

## Features

- Type-safe transfer methods: `mem_to_mem()`, `memset()`, `write()`, `read()`
- Scatter-gather builder for chained multi-segment transfers
- Async transfers with interrupt-driven completion
- Supports `u8`, `u16`, `u32` transfer widths

## Examples Verified on Hardware

All 10 examples tested on FRDM-MCXA276:
- Memory transfers: `dma_mem_to_mem`, `dma_memset`
- Scatter-gather: `dma_scatter_gather`, `dma_scatter_gather_builder`
- Advanced modes: `dma_ping_pong_transfer`, `dma_interleave_transfer`, `dma_wrap_transfer`, `dma_channel_link`
- UART with DMA: `lpuart_dma_tx`, `lpuart_ring_buffer_dma`

Fixes #33

bogdan-petru committed:
feat(dma): add DMA driver with 10 verified examples

Initial DMA driver implementation for MCXA276 with:

Core DMA Features:
- DmaChannel type with ownership tracking via Channel trait
- Transfer, RingBuffer, and ScatterGatherBuilder abstractions
- Support for mem-to-mem, mem-to-peripheral, peripheral-to-mem transfers
- Interrupt-driven completion with embassy async/await integration
- Word size abstraction (u8, u16, u32) via Word trait

LPUART DMA Integration:
- LpuartTxDma and LpuartRxDma drivers for async UART with DMA
- LpuartDma combined TX/RX driver
- Automatic chunking for buffers > 0x7FFF bytes
- DMA guards with Drop impl for safe cancellation

10 Verified Examples:
- dma_mem2mem: Basic memory-to-memory copy
- dma_memset: Memory fill with pattern
- dma_uart_tx: UART transmit via DMA
- dma_uart_rx: UART receive via DMA
- dma_uart_loopback: Combined TX/RX loopback test
- dma_scatter_gather: Linked descriptor chains
- dma_channel_linking: Major/minor loop channel linking
- dma_ring_buffer: Circular buffer for continuous streaming
- dma_ping_pong: Double-buffering pattern
- dma_software_trigger: Manual transfer triggering

PR Feedback Addressed:
- Use PAC accessor for LPUART DATA register instead of manual offset
- Add EnableInterrupt enum to replace boolean parameter for readability
- Add DMA guards with Drop impl for safe async cancellation
- Automatic chunking for large buffers instead of returning error
- Use NonNull<[W]> + PhantomData for RingBuffer (DMA acts like separate thread)
- Remove edma parameter from all methods (single eDMA instance steals ptr internally)
- Make edma_tcd() non-public (HAL should not expose PAC items)

jamesmunns commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+            return Err(Error::InvalidArgument);
+        }
+
+        let peri_addr = (self.info.regs as *const _ as *mut u8).wrapping_add(LPUART_DATA_OFFSET);
```
```suggestion
        let peri_addr = self.info.regs.data().as_ptr().cast::<u8>();
```

bogdan-petru replied:
done, using self.info.regs.data().as_ptr() now

jamesmunns commented on src/dma.rs:
```diff
@@ -0,0 +1,2482 @@
+        edma: &pac::edma_0_tcd0::RegisterBlock,
+        buf: &[W],
+        peri_addr: *mut W,
+        enable_interrupt: bool,
```
Consider making a simple enum here, like `enum EnableInterrupt { Yes, No }` so it makes calls easier to understand, like:

```rust
self.tx_dma.setup_write_to_peripheral(
    edma, buf, peri_addr, true, // Enable interrupt
);
// vs
self.tx_dma.setup_write_to_peripheral(
    edma, buf, peri_addr, EnableInterupt::Yes,
);
```

bogdan-petru replied:
ok, added EnableInterrupt enum with Yes/No variants

jamesmunns commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+        }
+
+        let len = buf.len();
+        if len > 0x7FFF {
```
It's probably unlikely, but it might be worth handling this internally, breaking up buf using `.chunks(0x7FFF)` and making repeated calls to an internal function, like:

```rust
let mut ttl = 0;
for chunk in buf.chunks(0x7FFF) {
  ttl += self.write_dma_inner(edma, request_source, chunk)?;
}

Ok(ttl)
```

This also automatically handles the case where `buf.len()` is zero.

bogdan-petru replied:
write_dma and read_dma now use .chunks() internally with write_dma_inner/read_dma_inner helpers.

jamesmunns commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+        .await;
+
+        // Cleanup
+        self.info.regs.baud().modify(|_, w| w.tdmae().disabled());
```
I'll make a comment elsewhere, but we need to use a more structured approach to this to ensure that the DMA transfer is dropped if the future is cancelled.

Right now, if we have a timeout, and abort the transfer without polling it to completion, the DMA will continue even though we have begun reusing the buffer/stack space.

This is usually by having a DMA method that returns a struct on transfer, and if that struct is dropped, it automatically stops the DMA.

bogdan-petru replied:
Transfer's Drop impl now aborts the DMA when the future is cancelled, so we don't continue writing to freed buffers.

jamesmunns commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+        }
+
+        let len = buf.len();
+        if len > 0x7FFF {
```
Same comment re: chunks

bogdan-petru replied:
Fixed in both write_dma and read_dma.

jamesmunns commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+            return Err(Error::InvalidArgument);
+        }
+
+        let peri_addr = (self.info.regs as *const _ as *const u8).wrapping_add(LPUART_DATA_OFFSET);
```
same comment re: `self.info.regs.data().as_ptr().cast::<u8>()`

bogdan-petru replied:
Fixed in both TX and RX paths.

jamesmunns commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+        cortex_m::asm::dsb();
+
+        // Cleanup
+        self.info.regs.baud().modify(|_, w| w.rdmae().disabled());
```
Same note about cancellation safety

bogdan-petru replied:
now using Transfer with Drop-based abort

jamesmunns commented on src/lpuart/mod.rs:
```diff
@@ -686,18 +675,38 @@ pub struct Lpuart<'a, M: Mode> {
+
+/// Lpuart driver with DMA support for both TX and RX.
+pub struct LpuartDma<'a, TxC: DmaChannelTrait, RxC: DmaChannelTrait> {
+    #[allow(dead_code)]
```
We can probably remove this if it is dead code.

bogdan-petru replied:
removed

jamesmunns commented on examples/src/bin/lpuart_ring_buffer.rs:
```diff
@@ -0,0 +1,162 @@
+    defmt::info!("LPUART Ring Buffer DMA example starting...");
+
+    // Enable DMA0 clock and release reset
+    unsafe {
```
These should be handled automatically in the HAL init, see how other peripherals like UART and I2C do this now.

bogdan-petru replied:
I agree, DMA clock/reset/init is now handled automatically by the HAL via ensure_init().

jamesmunns commented on examples/src/bin/lpuart_ring_buffer.rs:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    // Enable DMA interrupts
+    unsafe {
```
This should also be handled automatically in init, see how GPIO does this.

bogdan-petru replied:
NVIC enable remains explicit in example consistent with Embassy's pattern where interrupt binding is user-controlled via bind_interrupts!.

jamesmunns replied:
Can you explain what you mean? It's not usual that you need to explicitly unmask interrupts at the NVIC level for DMA to operate in embassy HALs. Can you link me to something that explains what you mean?

jamesmunns replied:
Okay, I think I understand this a bit better now, you're generally:

* Setting up the `Blocking` mode to get the UART initialized
* Then using low level DMA interfaces to get the hardware to do DMA driven reception

I think this is a reasonable low level demonstration that the DMA components work, though usually the examples included in the HAL repo are intended to show how we *want* users to use the HAL, and this would be generally non-typical.

CC @felipebalbi, can you think of where we would want to keep low level demos like this, if not the `examples` folder?

I also didn't realize at first why I was only getting data every 32 bytes, this is because interrupts only trigger on the half/full reception levels. The "continuous reception" part threw me off a bit, the hardware is continuous, but it's "chunky" in userspace.

bogdan-petru replied:
Refactored the example to use the high-level LpuartDmaRx API:

- Uses LpuartDmaRx::new() instead of raw DmaChannel::new()

- Uses setup_ring_buffer() which encapsulates all DMA/LPUART configuration

- Uses enable_dma_request() to start the transfer

- All PAC interactions are now hidden inside the HAL


Regarding chunky reception, yes, the half-transfer and complete-transfer interrupts are the only wakeup points with the current implementation.

jamesmunns commented on examples/src/bin/lpuart_ring_buffer.rs:
```diff
@@ -0,0 +1,162 @@
+    let pac_periphs = unsafe { pac::Peripherals::steal() };
+
+    // Initialize DMA
+    unsafe {
```
Should also be handled in the HAL init

bogdan-petru replied:
DMA init is now automatic in the HAL.

jamesmunns commented on examples/src/bin/lpuart_dma.rs:
```diff
@@ -0,0 +1,127 @@
+
+    defmt::info!("LPUART DMA example starting...");
+
+    // Enable DMA0 clock and release reset
```
See comments in `lpuart_ring_buffer` - all of this unsafe setup should't be handled in examples or user code.

bogdan-petru replied:
I agree

jamesmunns commented on examples/src/bin/lpuart_dma.rs:
```diff
@@ -0,0 +1,127 @@
+        dma::init(&pac_periphs);
+    }
+
+    // Get EDMA TCD register block for transfers
```
This should be handled automatically inside the lpuart driver

bogdan-petru replied:
Yes, request source is now derived automatically from the LPUART instance type via TX_DMA_REQ/RX_DMA_REQ.

jamesmunns commented on examples/src/bin/lpuart_dma.rs:
```diff
@@ -0,0 +1,127 @@
+
+    defmt::info!("Example complete");
+
+    loop {
```
You can remove this, returning from async main already does this.

bogdan-petru replied:
Removed

jamesmunns reviewed (changes requested):
Initial feedback, mostly looking at everything other than `src/dma.rs` and `examples/src/bin/dma_*.rs`.

I'll do a separate review for those files.

jamesmunns commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+            // Configure TCD for memory-to-peripheral transfer
+            // Using setup_write_to_peripheral which configures the TCD without returning
+            // a Transfer (which has a Drop impl that aborts). This allows us to use our
+            // own poll_fn-based waiting mechanism.
```
Could you comment on why you went with your own poll_fn-based approach, instead of doing something like `self.tx_dma.start_transfer(...).await;`? It seems like that would handle the concerns I mentioned re: cancellation safety.

bogdan-petru replied:
Refactored to use Transfer::start_transfer().await - cancellation safety is handled by Transfer's Drop impl which aborts the DMA.

jamesmunns commented on src/dma.rs:
```diff
@@ -0,0 +1,2482 @@
+    t.ch_int().write(|w| w.int().clear_bit_by_one());
+
+    // If DONE is set, this is a complete-transfer interrupt
+    let done = t.ch_csr().read().done().bit_is_set();
```
Did you intend to do something different for full/half interrupts here (maybe `half_waker` instead of `waker` twice)? Right now, you call the same thing regardless of whether we're done or not.

bogdan-petru replied:
You're right, that was a bug. Removed the else branch - waker now only fires when DONE is set.

jamesmunns commented on src/dma.rs:
````diff
@@ -0,0 +1,2482 @@
+/// ```
+pub struct RingBuffer<'a, W: Word> {
+    channel: AnyChannel,
+    buf: &'a mut [W],
````
```suggestion
    buf: NonNull<[W]>,
    _lt: PhantomData<&'a mut [W]>,
```

We *probably* want to replace storing this as a `&'a mut [W]` with something like `NonNull<[W]>`. In Rust, `&mut` is an "exclusive" reference, which the compiler is allowed to assume there will not be any "spooky" writes to it, which we are very clearly doing with DMA.

In embedded, we model DMA [as if they were a separate thread](https://onevariable.com/blog/interrupts-is-threads/), so we don't want to claim exclusive access when we've actually given our buffer to another "thread" to write to.

bogdan-petru replied:
Fixed - switched to NonNull<[W]> with PhantomData to avoid the exclusive reference issue.

jamesmunns commented on src/dma.rs:
```diff
@@ -0,0 +1,2482 @@
+    pub(crate) unsafe fn new(channel: AnyChannel, buf: &'a mut [W]) -> Self {
+        Self {
+            channel,
+            buf,
```
```suggestion
            buf: NonNull::from(buf),
            _lt: PhantomData,
```

bogdan-petru replied:
Applied

jamesmunns commented on src/dma.rs:
```diff
@@ -0,0 +1,2482 @@
+
+    /// Get a reference to the TCD register block for this channel.
+    #[inline]
+    fn tcd<'a>(&self, edma: &'a pac::edma_0_tcd0::RegisterBlock) -> &'a pac::edma_0_tcd0::Tcd {
```
Is there a reason to take the edma ptr here, instead of just stealing it like we do in `is_done`? I'm not sure if this adds any value.

bogdan-petru replied:
Removed the parameter - now using internal PAC access like is_done.

jamesmunns commented on src/dma.rs:
```diff
@@ -0,0 +1,2482 @@
+
+    /// Return a reference to the underlying TCD register block.
+    #[inline]
+    pub fn tcd<'a>(&self, edma: &'a pac::edma_0_tcd0::RegisterBlock) -> &'a pac::edma_0_tcd0::Tcd {
```
Same note re: edma register block

bogdan-petru replied:
Fixed.

jamesmunns commented on src/dma.rs:
```diff
@@ -0,0 +1,2482 @@
+    /// - The pattern and destination buffer must remain valid for the duration of the transfer.
+    pub unsafe fn memset_with_edma<W: Word>(
+        &self,
+        edma: &pac::edma_0_tcd0::RegisterBlock,
```
The system only has a single edma instance, right?

bogdan-petru replied:
Correct, MCXA276 has a single eDMA instance. We now access it internally rather than passing it around.

jamesmunns commented on src/dma.rs:
```diff
@@ -0,0 +1,2482 @@
+/// This is safe to call at any time, but the caller must ensure that
+/// concurrent access to the same DMA channel is properly synchronized.
+#[inline]
+pub fn edma_tcd() -> &'static pac::edma_0_tcd0::RegisterBlock {
```
Ideally this isn't necessary to expose as `pub` - we generally don't show PAC items at the HAL level

bogdan-petru replied:
Removed edma_tcd() - it's no longer exposed.

jamesmunns reviewed (commented):
Some more comments on the `dma.rs` file, I haven't really looked at the `dma_*` examples yet.

felipebalbi commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+    pub fn new<T: Instance>(
+        _inner: Peri<'a, T>,
+        tx_pin: Peri<'a, impl TxPin<T>>,
+        tx_dma_ch: Peri<'a, C>,
```
Is it so that any dma channel can be used for any uart instance? If the hardware has a static mapping of channels, then that needs to be encoded here as well

bogdan-petru replied:
MCXA276 uses a flexible DMA mux - any channel can connect to any peripheral. Request source is now derived from the Instance trait.

felipebalbi commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+        tx_pin.as_tx();
+        let tx_pin: Peri<'a, AnyPin> = tx_pin.into();
+
+        Lpuart::<Blocking>::init::<T>(Some(&tx_pin), None, None, None, config)?;
```
Should be `Async`, right?

bogdan-petru replied:
yes, using Lpuart::<Async>::init now.

felipebalbi commented on src/lpuart/mod.rs:
```diff
@@ -1019,10 +1026,283 @@ impl<'a> Lpuart<'a, Blocking> {
+    /// * `edma` - Reference to the EDMA TCD register block
+    /// * `request_source` - DMA request source number (e.g., `dma::DMA_REQ_LPUART2_TX`)
+    /// * `buf` - Data buffer to transmit
+    pub async fn write_dma(&mut self, edma: &EdmaRegs, request_source: u8, buf: &[u8]) -> Result<usize> {
```
Not sure exposing raw register access to the dma is a good idea.

bogdan-petru replied:
Removed - DMA register access is now internal to the driver

felipebalbi commented on src/dma.rs:
```diff
@@ -0,0 +1,2482 @@
+pub const DMA_REQ_LPUART4_RX: u8 = 29;
+pub const DMA_REQ_LPUART4_TX: u8 = 30;
+pub const DMA_REQ_LPUART5_RX: u8 = 31;
+pub const DMA_REQ_LPUART5_TX: u8 = 32;
```
ok, so there are static mappings. Have a look at how this is usually implemented [here](https://github.com/embassy-rs/embassy/blob/main/embassy-imxrt/src/flexcomm/uart.rs#L1200-L1230)

bogdan-petru replied:
MCXA276 uses INPUTMUX for flexible DMA routing rather than static channel mappings - any channel can connect to any peripheral. We derive the request source from TX_DMA_REQ/RX_DMA_REQ in the Instance trait, similar concept to the i.MX RT pattern but adapted for the flexible mux architecture.

bogdan-petru committed:
Updated per PR #52 feedback

bogdan-petru commented:
Addressed the feedback. Removed PAC exposure (edma_tcd gone, EdmaRegs params removed from write_dma/read_dma), DMA now auto-initializes via ensure_init(), and request_source is derived from the LPUART instance type. Fixed on_interrupt to properly wake half_waker vs waker. All examples updated to use bind_interrupts! macro.

bogdan-petru committed:
Apply rustfmt formatting fixes

jamesmunns commented on examples/src/bin/lpuart_ring_buffer.rs:
```diff
@@ -0,0 +1,149 @@
+    tx.blocking_write(b"==============================\r\n\r\n").unwrap();
+
+    // Get LPUART2 RX data register address for DMA
+    let lpuart2 = unsafe { &*pac::Lpuart2::ptr() };
```
Does the current ring buffer API require users to do direct PAC interactions?

IMO we should handle this internal to the HAL if at all possible. It is very unusual to require unsafe for pac-level interactions for high-level HAL APIs.

bogdan-petru replied:
The example no longer requires any PAC-level access.

Added LpuartRxDma::setup_ring_buffer() and enable_dma_request() methods that encapsulate:

- DMA request source configuration

- RDMAE (RX DMA Enable) bit setting

- Circular DMA transfer setup

- NVIC interrupt enabling (via DmaChannel::enable_interrupt())

The only remaining unsafe is for the static buffer access.

felipebalbi commented on src/lpuart/mod.rs:
```diff
@@ -90,13 +82,29 @@ macro_rules! impl_instance {
                         = crate::clocks::periph_helpers::LpuartInstance::[<Lpuart $n>];
                     type Interrupt = crate::interrupt::typelevel::[<LPUART $n>];
+                    const TX_DMA_REQ: u8 = $tx_req;
+                    const RX_DMA_REQ: u8 = $rx_req;
```
DMA request lines should be defined as `Peri` too. Not raw ints.

bogdan-petru replied:
Implemented a type-safe DMA request source API:

- Added DmaRequest trait with sealed pattern in  src/dma.rs

- Created Lpuart{0-5}{Tx,Rx}Request marker types for each LPUART DMA request

- Updated DmaChannel::set_request_source() to accept R: DmaRequest generic

- Replaced Instance::TX_DMA_REQ: u8 / RX_DMA_REQ: u8 constants with associated types TxDmaRequest / RxDmaRequest

felipebalbi commented on src/lpuart/mod.rs:
```diff
@@ -1027,10 +1053,377 @@ impl<'a> Lpuart<'a, Blocking> {
+// ============================================================================
+
+/// Maximum bytes per DMA transfer (eDMA CITER/BITER are 15-bit fields).
+const DMA_MAX_TRANSFER_SIZE: usize = 0x7FFF;
```
is this a property of the UART or of the DMA?

bogdan-petru replied:
It's a property of the DMA controller. Moved to src/dma.rs and imported in lpuart via use crate::dma::DMA_MAX_TRANSFER_SIZE .

felipebalbi commented on src/dma.rs:
```diff
@@ -0,0 +1,2495 @@
+            });
+        }
+    }
+}
```
how about adding an `init` called that's called when the HAL is initialized?

bogdan-petru replied:
Added a public dma::init() function that is now called from hal::init() alongside gpio::init(). Also removed the ensure_init() function and its calls from DmaChannel::new() and examples.

felipebalbi commented on src/lib.rs:
```diff
@@ -51,6 +52,14 @@ embassy_hal_internal::peripherals!(
 
     DBGMAILBOX,
     DMA0,
+    DMA_CH0,
+    DMA_CH1,
+    DMA_CH2,
+    DMA_CH3,
+    DMA_CH4,
+    DMA_CH5,
+    DMA_CH6,
+    DMA_CH7,
```
oh, you have them as peripherals, in that case impl the uart dma trait for the peripheral.

bogdan-petru replied:
Are you suggesting something like:

pub trait UartDma {
    type TxChannel: DmaChannel;
    type RxChannel: DmaChannel;
    fn tx_dma_request() -> DmaRequest;
    fn rx_dma_request() -> DmaRequest;
}

bogdan-petru committed:
refactor(dma): move DMA_MAX_TRANSFER_SIZE to dma module and init during HAL startup

Address felipebalbi's review comments on PR #52:

- Move DMA_MAX_TRANSFER_SIZE constant from lpuart/mod.rs to dma.rs
  where it logically belongs (describes eDMA4 hardware limitation)
- Add public dma::init() function called during hal::init() instead
  of lazy initialization via ensure_init()
- Remove ensure_init() entirely since it's no longer needed
- Remove ensure_init() calls from DmaChannel::new() and from_token()
- Remove ensure_init() calls from examples (dma_channel_link, dma_scatter_gather)
- Refactor lpuart_ring_buffer example to use LpuartDma::new() + split()
  pattern instead of separate TX/RX drivers
- Add [lints.rust] section to suppress unexpected_cfgs warning for 'rt'
  feature used by embassy_hal_internal::interrupt_mod! macro

This makes DMA initialization explicit during HAL startup (like GPIO)
and keeps DMA-specific constants in the DMA module.

bogdan-petru committed:
feat(dma): implement type-safe DMA request source API

Replace raw u8 DMA request sources with typed request system
following the embassy-imxrt pattern. This provides compile-time
verification that peripherals use correct DMA request sources.

Changes:
- Define DmaRequest trait for type-safe request sources
- Add sealed SealedDmaRequest trait to prevent external implementation
- Create Lpuart{0-5}{Tx,Rx}Request marker types for each LPUART DMA request
- Update DmaChannel::set_request_source() to accept generic R: DmaRequest
- Replace Instance::TX_DMA_REQ/RX_DMA_REQ u8 constants with associated types
- Update all LPUART DMA call sites to use typed request API

This prevents errors like using LPUART2_RX request for LPUART0
at compile time with zero runtime overhead.

Addresses felipebalbi's review comment on PR #52 line 86.

bogdan-petru committed:
Merge upstream/main to incorporate LPUART constructor refactor (#51) and GPIO Drive/Slew fix (#57)

jamesmunns commented:
Hey @bogdan-petru could you make your fork public? It looks like it remained private when the main repo went public.
